cpp/src/arrow/compute/exec/bloom_filter.cc

```diff
@@ -0,0 +1,434 @@
+
+  thread_local_states_.resize(num_threads);
+  prtn_locks_.Init(1 << log_num_prtns_);
+
+  RETURN_NOT_OK(build_target->CreateEmpty(num_rows, pool));
+
+  return Status::OK();
+}
+
+Status BloomFilterBuilder_Parallel::PushNextBatch(size_t thread_id, int num_rows,
+                                                  const uint32_t* hashes) {
+  PushNextBatchImp(thread_id, num_rows, hashes);
+  return Status::OK();
+}
+
+Status BloomFilterBuilder_Parallel::PushNextBatch(size_t thread_id, int num_rows,
+                                                  const uint64_t* hashes) {
+  PushNextBatchImp(thread_id, num_rows, hashes);
+  return Status::OK();
+}
+
+template <typename T>
+void BloomFilterBuilder_Parallel::PushNextBatchImp(size_t thread_id, int num_rows,
+                                                   const T* hashes) {
+  int num_prtns = 1 << log_num_prtns_;
+  ThreadLocalState& local_state = thread_local_states_[thread_id];
+  local_state.partition_ranges.resize(num_prtns + 1);
+  local_state.partitioned_hashes_64.resize(num_rows);
+  local_state.unprocessed_partition_ids.resize(num_prtns);
+  uint16_t* partition_ranges = local_state.partition_ranges.data();
+  uint64_t* partitioned_hashes = local_state.partitioned_hashes_64.data();
+  int* unprocessed_partition_ids = local_state.unprocessed_partition_ids.data();
+
+  PartitionSort::Eval(
+      num_rows, num_prtns, partition_ranges,
+      [hashes, num_prtns](int row_id) {
+        constexpr int kLogBlocksKeptTogether = 7;
+        constexpr int kPrtnIdBitOffset =
+            BloomFilterMasks::kLogNumMasks + 6 + kLogBlocksKeptTogether;
+        return (hashes[row_id] >> (kPrtnIdBitOffset)) & (num_prtns - 1);
+      },
+      [hashes, partitioned_hashes](int row_id, int output_pos) {
+        partitioned_hashes[output_pos] = hashes[row_id];
+      });
+
+  int num_unprocessed_partitions = 0;
+  for (int i = 0; i < num_prtns; ++i) {
+    bool is_prtn_empty = (partition_ranges[i + 1] == partition_ranges[i]);
+    if (!is_prtn_empty) {
+      unprocessed_partition_ids[num_unprocessed_partitions++] = i;
+    }
+  }
+  while (num_unprocessed_partitions > 0) {
+    int locked_prtn_id;
+    int locked_prtn_id_pos;
+    prtn_locks_.AcquirePartitionLock(num_unprocessed_partitions,
+                                     unprocessed_partition_ids,
+                                     /*limit_retries=*/false, /*max_retries=*/-1,
+                                     &locked_prtn_id, &locked_prtn_id_pos);
+    build_target_->Insert(
+        hardware_flags_,
+        partition_ranges[locked_prtn_id + 1] - partition_ranges[locked_prtn_id],
+        partitioned_hashes + partition_ranges[locked_prtn_id]);
+    prtn_locks_.ReleasePartitionLock(locked_prtn_id);
+    if (locked_prtn_id_pos < num_unprocessed_partitions - 1) {
+      unprocessed_partition_ids[locked_prtn_id_pos] =
+          unprocessed_partition_ids[num_unprocessed_partitions - 1];
+    }
+    --num_unprocessed_partitions;
+  }
+}
```

Comment:
So each thread will keep picking partitions at random until all partitions are gone?  It's a sort of dynamic scheduling approach that handles load balancing automatically?

I think that is clever (as we won't really know in general at any point in time how many threads are available to do work) and I'm curious to see how it would be used in practice.

Reply:
Yes, the idea is that a thread has to visit multiple partitions but each partition is shared. Instead of taking mutexes on partitions in a fixed order (which would be unacceptably slow) threads visit partitions at random and if one partition is already occupied they move to a different one. So as long as there are enough partitions compared to threads, threads should not collide often. I also verified on my machine with reasonably high number of CPU cores that this works well in practice. Plus it simplifies some parallel algorithms implementation.